### PR TITLE
Giftcard 2.0 - Recipient fields

### PR DESCRIPTION
**Pull request by philbirt**
* Add recipient-related fields to gift cards
* Ability to add, remove, update cart with gift card meta data
* Admin for editing gift cards

### CODE REVIEW

**Comment by athal7** [comment]
looking good

**Comment by philbirt** [comment]
Alright this is now stable and finalized. Mind taking one last look @athal7 ?

File: app/models/concerns/spree/gift_cards/order_concerns.rb [hunk 1]
```diff
@@ -11,17 +11,22 @@ module GiftCards::OrderConcerns
     end
 
     module InstanceMethods
-      def finalize!
-        create_gift_cards
-        super
+      def gift_card_match(line_item, options)
+        gift_card_options = options["gift_card_details"]
+        !line_item.gift_card? ||
+          (line_item.gift_card? && gift_card_options.present? &&
```
Comment (athal7) [inline app/models/concerns/spree/gift_cards/order_concerns.rb]: would this be able to be simplified as:

```ruby
  return true unless line_item.gift_card?
  line_item.gift_cards.any? { |gift_card| gift_card.details == options["gift_card_details"] }
```

if you push https://github.com/solidusio/solidus_virtual_gift_card/pull/9/files#diff-94036a0cdf611599347aed2b6e700caaR18 down to the gift card model?

File: app/models/concerns/spree/gift_cards/order_concerns.rb [hunk 2]
```diff
@@ -11,17 +11,22 @@ module GiftCards::OrderConcerns
...
+          line_item.gift_cards.any? {|gc| gc.gift_message == gift_card_options['gift_message']})
       end
 
-      def create_gift_cards
-        line_items.each do |item|
-          item.quantity.times do
-            Spree::VirtualGiftCard.create!(amount: item.price, currency: item.currency, purchaser: user, line_item: item) if item.gift_card?
-          end
+      def finalize!
+        gift_cards.each do |gift_card|
```
Comment (athal7) [inline app/models/concerns/spree/gift_cards/order_concerns.rb]: what happens if one of the later order finalization steps fails? seems to me this should go after the super behavior, not before

File: app/models/concerns/spree/gift_cards/order_contents_concerns.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  module GiftCards::OrderContentsConcerns
+    extend ActiveSupport::Concern
+
+    included do
+      prepend(InstanceMethods)
+    end
+
+    module InstanceMethods
+      def add(variant, quantity = 1, options = {})
+        line_item = super
```
Comment (athal7) [inline app/models/concerns/spree/gift_cards/order_contents_concerns.rb]: don't take this as real feedback, just a thought if you wanted this is also an option:

```ruby
super.tap do |line_item|
  create_gift_cards(line_item, quantity, options["gift_card_details"] || {})
end
```

File: app/models/concerns/spree/gift_cards/order_contents_concerns.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+              purchaser_name: gift_card_details["purchaser_name"],
+              gift_message: gift_card_details["gift_message"],
+            )
+          end
+        end
+      end
+
+      def remove_gift_cards(line_item, quantity_diff)
+        if line_item.gift_card?
+          line_item.gift_cards.take(quantity_diff).each(&:destroy)
```
Comment (athal7) [inline app/models/concerns/spree/gift_cards/order_contents_concerns.rb]: do we want to be explicit about the ordering here in any way?

**Comment by philbirt** [comment]
Excellent reviewing @athal7. The gift_card_match simplification wasn't quite as simple as hoped, but I got to something a bit more concise. Great catch on all that. One more :eyes: ?

File: app/models/concerns/spree/gift_cards/order_contents_concerns.rb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+              purchaser_name: gift_card_details["purchaser_name"],
+              gift_message: gift_card_details["gift_message"],
+            )
+          end
+        end
+      end
+
+      def remove_gift_cards(line_item, quantity_diff)
+        if line_item.gift_card?
+          line_item.gift_cards.order(:created_at).last(quantity_diff).each(&:destroy)
```
Comment (athal7) [inline app/models/concerns/spree/gift_cards/order_contents_concerns.rb]: do we want destroy!. seems consistent with what we are doing for create

**Comment by athal7** [comment]
:+1:

**Comment by jordan-brough** [comment]
Merging on behalf of @athal7 via @philbirt.

**Pull request merged**